demos/link-prediction/random-walks/ctdne-link-prediction.ipynb

```diff
@@ -0,0 +1,686 @@
+{
```

Comment:
<p>Now that https://github.com/stellargraph/stellargraph/pull/1033 is merged, this can be slightly simplified potentially, although maybe we want to be sure we've got all the nodes using the main graph? (Also the <code>source_column</code> and <code>target_column</code> parameters are being set to their default and so don't seem necessary?)</p><p> </p><pre class="ql-syntax" spellcheck="false">graph = StellarGraph(edges=edges_graph, edge_weight_column="time")
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/stellargraph/stellargraph/pull/1056/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Reply:
Hmm it seems like the reduced graph has 2 less nodes than the full graph actually:
```
Number of nodes in full graph: 151
Number of nodes in training graph: 149
```
this is handled already for temporal walks by the function that grabs the embedding with `try: ... except KeyError`, but I'd have to add the same logic for static walks too. Alternatively:
* keep passing in `nodes=pd.DataFrame(index=full_graph.nodes())` when constructing the graph
* update the dataset loader to make sure every node still exists in the reduced graph

Reply:
@kjun9  whats the reduced graph? I might have missed something.

Reply:
The reduced graph is from taking the first 75% of edges from the full dataset as the "graph", and the remaining 25% edges for training and test examples

Reply:
Is this how they are doing it in the paper?

Reply:
But yeah this would make sense that some nodes might be left out with the edge splitting.

Reply:
Yep, from the paper:
> To generate a set of labeled examples for link prediction, we first sort the edges in each graph by time (ascending) and use the first 75% for representation learning. The remaining 25% are considered as positive links and we sample an equal number of negative edges randomly

Reply:
Another detail that the paper brushed under the rug...
Your alternative sounds more proper to me. With the note in the demo that the original method does not explain in their experimental section how they handle the missing nodes after the split and this is how you are taking care of it.

Reply:
> keep passing in nodes=pd.DataFrame(index=full_graph.nodes()) when constructing the graph

Given reducing the code would likely break it, going with this option and keeping it as is seems fine to me.